src/liballoc/collections/vec_deque.rs

```diff
@@ -1834,8 +1834,144 @@ impl<T> VecDeque<T> {
     #[inline]
     #[stable(feature = "append", since = "1.4.0")]
     pub fn append(&mut self, other: &mut Self) {
-        // naive impl
-        self.extend(other.drain(..));
+        // Copy from src[i1..i1 + len] to dst[i2..i2 + len].
+        // Does not check if the ranges are valid.
+        unsafe fn copy_part<T>(i1: usize, i2: usize, len: usize, src: &[T], dst: &mut [T]) {
+            debug_assert!(src.get(i1..i1 + len).is_some() && dst.get(i2..i2 + len).is_some());
+            ptr::copy_nonoverlapping(src.as_ptr().add(i1), dst.as_mut_ptr().add(i2), len);
+        }
+
+        let src_total = other.len();
+
+        // Guarantees there is space in `self` for `other`.
+        self.reserve(src_total);
+
+        self.head = {
+            let dst_start_1 = self.head;
+            let src_start_1 = other.tail;
+            let dst_wrap_point = self.cap();
+            let src_wrap_point = other.cap();
+
+            let dst = unsafe { self.buffer_as_mut_slice() };
+            let src = unsafe { other.buffer_as_slice() };
+
+            let src_wraps = other.tail > other.head;
+            let dst_wraps = dst_start_1 + src_total > dst_wrap_point;
+
+            // When minimizing the amount of calls to `copy_part`, there are
+            // 6 different cases to handle. Whether src and/or dst wrap are 4
+            // combinations and there are 3 distinct cases when they both wrap.
+            // 6 = 3 + 1 + 1 + 1
+            match (src_wraps, dst_wraps) {
```

Comment:
Pure speculation, but I wonder if there could be a method to encapsulate this logic, something along the lines of

```rust
fn source_parts(&self) -> (&[T], &[T]); // might exist as `as_slices`
unsafe fn destination_parts(&mut self) -> (&mut [T], &mut [T]);
``` 

This code will still need to exist and handle the same cases to minimize the copies.